# Solutions to precourse assignment
1. [Intro Jupyter Notebook](#1.-Intro-Jupyter-Notebook)
2. [Strings](#2.-Strings)
3. [Lists](#3.-Lists)
4. [Control Flow](#4.-Control-Flow)
5. [Functions](#5.-Functions)
6. [Tuples,  Sets and Dictionaries](#6.-Tuples,-Sets-and-Dictionaries)

# 1. Intro Jupyter Notebook

### Problem 1
<span style="color:green">How many seconds are there in a century?</span>

In [1]:
# your code here
60 * 60 * 24 * 365.25 * 100

3155760000.0

### Advanced: Problem 2
<span style="color:green">Write an expression using only operator and math libraries to see if the area of a circle with radius 5 is greater than or equal to 40</span>

In [5]:
# your code here
from operator import mul, ge
from math import pi
ge(mul(pow(5, 2), pi), 40) # pow is both a builtin function and part of the math library

True

### Problem 3
<span style="color:green">Use three different variables to store a 15% sales tax on a $70 meal to get the total and return them all as output at the same time</span>

In [6]:
# your code here
tax_rate = .15
meal = 70
total = (tax_rate + 1) * meal
tax_rate, meal, total

(0.15, 70, 80.5)

# 2. Strings

### Problem 1
<span style="color:green">Find and use the string method that will replace all each letter a with A. </span>

In [7]:
replace_string = 'replace each letter a in this string with A'
# your code here
replace_string.replace('a', 'A')

'replAce eAch letter A in this string with A'

### Problem 2
<span style="color:green">Strip all the letter a's from the right side, switch the case of each letter(ie from lower to upper and from upper to lower) and find the position(aka index) of the first letter 'o' </span>

In [11]:
test_string = 'aaaa TOO many aaaaaaaaa'
# your code here
test_string.rstrip('a').swapcase().index('o')

6

### Problem 3
<span style="color:green">Slice this string from index 5 to the end by every 4th element</span>

In [12]:
test_string = 'or is this precourse too difficult?'
# your code here
test_string[5::4]

' sertdit'

### Problem 4
<span style="color:green">Get every third element starting from the end to the beginning </span>

In [13]:
# your code here
test_string[::-3]

'?ufdoeue hsr'

### Problem 5
<span style="color:green">Use four chained operators on a string of your choice </span>

In [17]:
# Enter in a string inside the quotes
your_string = '        blah blah and more blah'
# your code here
your_string.title().ljust(0).strip().split() # returns list

['Blah', 'Blah', 'And', 'More', 'Blah']

# 3. Lists

### Problem 1
<span style="color:green">Mutate `list_2` and print out both lists. Does mutating list_2 still change list_1? </span> 

In [2]:
list_1 = [1, 5, 10]
list_2 = list_1

# your code here
list_2[0] = 'changed'

print(list_1) # yes, list_1 mutated
print(list_2) 

['changed', 5, 10]
['changed', 5, 10]


### Advanced: Problem 2
<span style="color:green">Create a variable that is assigned to a list of five lists where the inner lists have different numbers of elements and different types of elements. Use the sort method to sort the outer list by the number of elements in each inner list. If you are confused about list of lists, [click here](#Lists-of-Lists)</span> 

In [3]:
# your code here
list_list = [[87, 'asdf'], ['a', True, 56], [-90, 5j, 6], [False, 1, 2, 3, 4], [5]]
list_list.sort(key=len)
list_list

[[5], [87, 'asdf'], ['a', True, 56], [-90, 5j, 6], [False, 1, 2, 3, 4]]

### Problem 3
<span style="color:green">Notice above that 'greater than 10' was surrounded by [ ]. This means that ['greater than 10'] is a list. Do the same operation as the cell above except without the [ ]. What happens?</span> 

In [4]:
# your code here
my_list = list(range(20))
my_list[11:] = 'greater than 10'
my_list # every character is made into an element of the list

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 'g',
 'r',
 'e',
 'a',
 't',
 'e',
 'r',
 ' ',
 't',
 'h',
 'a',
 'n',
 ' ',
 '1',
 '0']

### Problem 4
<span style="color:green">Create a list of 10 elements starting from 0 using the range function. Slice the list from the 2nd element to the 6th element with a step size of 2 and attempt to assign a single element list to it. What happens?</span> 

In [7]:
# your code here
my_list = list(range(10))
my_list[2:6:2] = [5] # error tells us that slice is size of 2 and we tried to assign 1 element list

ValueError: attempt to assign sequence of size 1 to extended slice of size 2

### Problem 5
<span style="color:green">After the error above in problem 4, try and assign a different list to that slice that does not raise an error. What is the rule for assigning new elements to slices of lists?</span> 

In [10]:
# your code here
my_list[2:6:2] = [99, 100]
my_list # you must use a list the same length as the slice

[0, 1, 99, 3, 100, 5, 6, 7, 8, 9]

### Problem 6
<span style="color:green">Create a list using the range function from 100 to 500. Then create a slice object that starts from the 20th element from the end and slices to the end (use keyword `None` for this) by 5. Use this slice object in the list you created.</span> 

In [28]:
# your code here
my_list = list(range(100, 500))
my_slice = slice(-20, None, 5)
my_list[my_slice]

[480, 485, 490, 495]

### Problem 7
<span style="color:green">Create a list of characters using the list function on a string. Delete elements starting from index 2 up to index 6. Output the new list.
</span> 

In [29]:
# your code here
my_list = list('characters deleted')
del my_list[2:6]
my_list

['c', 'h', 't', 'e', 'r', 's', ' ', 'd', 'e', 'l', 'e', 't', 'e', 'd']

### Advanced: Problem 8
<span style="color:green">Create a list using the range function from 352 to 5218 by 7. Delete elements starting from the index that is 100th from the end up to the 8th element from the end by 6. Without printing out the entire list, what is the last number that was deleted from your original list? 
</span> 

In [31]:
# your code here
a = list(range(352,5218, 7))
del a[-100:-8:6]
a[-20:]

[5063,
 5077,
 5084,
 5091,
 5098,
 5105,
 5119,
 5126,
 5133,
 5140,
 5147,
 5161,
 5168,
 5175,
 5182,
 5189,
 5196,
 5203,
 5210,
 5217]

Looking at the last 20 elements we can see where there was an increment of 14 (and not 7) from one element to the next. We can see that the increment of 14 took place from 5147 to 5161 meaning that the element 5154 was the last element deleted.

### Problem 9
<span style="color:green">Use the character 'a' to separate and join the following list to reveal the magic word </span>

In [35]:
# concatenate all elements of this list together separated by 'a'
magic_word = ['', 'b', 'r', 'c', 'd', 'b', 'r', '!']

# your code here
'a'.join(magic_word)

'abaracadabara!'

### Problem 10
<span style="color:green">use the + operator to concatenate two lists. Make the first list the even elements indices of test_list and the second list the odd elements of test_list</span>

In [36]:
test_list = list(range(20))

# your code here
test_list[::2] + test_list[1::2]

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 1, 3, 5, 7, 9, 11, 13, 15, 17, 19]

### Problem 11
<span style="color:green">Create a list then use the `extend` method to add elements to it</span>

In [37]:
test_list = list('abcdefghijkl')
# your code here
test_list.extend(['more', 'elements'])
test_list

['a',
 'b',
 'c',
 'd',
 'e',
 'f',
 'g',
 'h',
 'i',
 'j',
 'k',
 'l',
 'more',
 'elements']

### Problem 12
<span style="color:green">In words, without actually programming, what will <strong>my_list * 5</strong> do?</span>

Enter you answer (double click here to edit cell):  
`my_list * 5` will concatenate itself to itself 5 times.

In [34]:
print(test_list * 5)

['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'more', 'elements', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'more', 'elements', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'more', 'elements', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'more', 'elements', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'more', 'elements']


### Problem 13
<span style="color:green">First create a list containing at least three inner lists. Then replace the second list with a reverse of the third list.</span>

In [35]:
# your code here
list_of_lists = [[1,2,3], [4,5,6], [7,8,9]]
list_of_lists[1] = list_of_lists[2][::-1]
list_of_lists

[[1, 2, 3], [9, 8, 7], [7, 8, 9]]

### Problem 14
<span style="color:green">Create a list using the range function from 0 to 100,000 by 113 and determine if the value 87,649 is in the list or not.</span>

In [48]:
# your code here
my_list = list(range(0, 100000, 113))
87649 in my_list

False

# 4. Control Flow

### Problem 1
<span style="color:green">Create a different string that will trigger the second print statement from above. </span> 

In [43]:
# your code here
test_string = 'asdfjhweroiuk' # change this string so the second print statement below is triggered

# trigger the second print statement
if test_string.count('a') > 4:
    print("There are more than 4 a's in your string")
elif test_string.find('k') > 10:
    print("The first k occurs after the 11th element")
else:
    print("My super intelligent responses did not find any info on your string. Please change it")

The first k occurs after the 11th element


### Problem 2
<span style="color:green">Create a different string that will trigger the third print statement from above. </span> 

In [1]:
# your code here
test_string = 'asdfjhweroiu' # change this string so the third print statement below is triggered

# trigger the third print statement
if test_string.count('a') > 4:
    print("There are more than 4 a's in your string")
elif test_string.find('k') > 10:
    print("The first k occurs after the 11th element")
else:
    print("My super intelligent responses did not find any info on your string. Please change it")

My super intelligent responses did not find any info on your string. Please change it


### Problem 3
<span style="color:green">Write an expression that returns true if an integer is either greater than 10 or divisible by 7.</span> 

In [45]:
# your code here
x = 40
(x > 10) or (x % 7 == 0)

True

### Problem 4
<span style="color:green">Write an expression that returns true if the last character of a string is not 's'</span> 

In [47]:
test_string = 'asfdaskfjsafl'
# your code here
test_string[-1] != 's'

True

### Problem 5
<span style="color:green">To get higher decimal precision you will need to use more points. Write a loop (without drawing any figures) that calculates the value of $\pi$ to 2 decimal places using the same simulation technique as above.</span> 

In [8]:
import random

In [4]:
# your code here
count = 0
n = 100000
for i in range(n):
    x = random.random()
    y = random.random()
    if x ** 2 + y ** 2 <= 1:
        count += 1
pi = 4 * count / n
print(pi)

3.14264


### Problem 6
<span style="color:green">Use a for loop to iterate over the first 10 positive integers, printing out their squared value</span>

In [48]:
#your code here
for i in range(1,11):
    print(i ** 2)

1
4
9
16
25
36
49
64
81
100


### Problem 7
<span style="color:green">Use a for loop to iterate over the first 10 positive integers, printing out their squared value only if the value is greater than 50</span>

In [5]:
#your code here
for i in range(1, 11):
    squared_value = i ** 2
    if squared_value > 50:
        print(squared_value)

64
81
100


### Problem 8
<span style="color:green">Use a for loop to iterate over every sixth integer of the first 1000, appending their squared value to a list. Do not print this result out.</span>

In [52]:
squared_list = []
# your code here
for i in range(1000)[::6]:
    squared_list.append(i ** 2)

### Problem 9
<span style="color:green">Use a while loop to find the sum of all integers between a given postive integer and 0. Print out result.</span>

In [3]:
x = 20
total = 0
# your code here
while x > 0:
    total += x
    x -= 1
print('Total is {}'.format(total))

Total is 210


### Problem 10
<span style="color:green">Use a while loop to find the sum of all integers between a given integer and its square root. Use math.sqrt to find the square root. </span>

In [5]:
import math
x = 20
total = 0
# your code here
sqrt = math.sqrt(x)
while x >= sqrt:
    total += x
    x -= 1
print('Total is {}'.format(total))

Total is 200


### Problem 11: Advanced
<span style="color:green">In this problem you will use a while loop to try and guess a randomly generated integer between 1 and 1000. The only information you will be given is whether your guess is higher or lower than the unknown integer. Your new guess will be updated in an optimal way. The optimal way to update your guess is to choose the next integer exactly half way between your current guess and the lower/upper boundary.

Steps:
1. Generate random integer
2. Generate first guess
3. Begin while loop
4. Check if guess is correct
5. Change lower and upper boundaries of guessing region based on if guess if higher/lower than unknown number
6. Keep track of number of guesses
7. Print out upper and lower boundary, the current guess and the unknown number at each iteration to track progress
8. Print out one last statement outside of the while loop that shows the guess is actually correct and the number of guesses that it took to get there.

Tips:
1. Make sure to guess is an integer. Use `int` function.
</span>

In [32]:
upper = 1000
lower = 1
unknown = random.randint(1, 1000) #generate unknown random integer
guess = int((upper + lower) / 2)
num_guesses = 1
#  your code here
while guess != unknown:
    print('Guess was {}, lower boundary is {}, upper boundary is {}, unknown is {}'.format(guess, lower, upper, unknown))
    if guess > unknown:
        upper = guess - 1
    else:
        lower = guess + 1
    guess = int((upper + lower) / 2)
    num_guesses += 1

print('Got it in {} guesses! Guess was {} and Unknown was {}. Final region was ({}, {})'.format(num_guesses, guess, unknown, lower, upper)) 

Guess was 500, lower boundary is 1, upper boundary is 1000, unknown is 426
Guess was 250, lower boundary is 1, upper boundary is 499, unknown is 426
Guess was 375, lower boundary is 251, upper boundary is 499, unknown is 426
Guess was 437, lower boundary is 376, upper boundary is 499, unknown is 426
Guess was 406, lower boundary is 376, upper boundary is 436, unknown is 426
Guess was 421, lower boundary is 407, upper boundary is 436, unknown is 426
Guess was 429, lower boundary is 422, upper boundary is 436, unknown is 426
Guess was 425, lower boundary is 422, upper boundary is 428, unknown is 426
Guess was 427, lower boundary is 426, upper boundary is 428, unknown is 426
Got it in 10 guesses! Guess was 426 and Unknown was 426. Final region was (426, 426)


### Problem 12
<span style="color:green">Use the `while true` logic to continually generate the roll of two 6-sided dice. Stop the loop with a break statement when the total of the current roll is 7 or 11. Use `random.randint(1, 6)` to roll a single die. Print out the number of rolls it took to get a 7 or 11. </span>

In [81]:
# your code here
n = 0
while True:
    n += 1
    die_1 = random.randint(1, 6)
    die_2 = random.randint(1, 6)
    if die_1 + die_2 in [7, 11]:
        break
print("It took {} rolls to get 7 or 11".format(n))

### Problem 13
<span style="color:green">Create 100 trials, where each trial is the number of times it takes to roll two dice that sum to 7 or 11. Keep track of the result of each trial in a list. You will need nested loops for this. Then find the average number of rolls it takes using the `sum` and `len` functions on your list</span>

In [82]:
# your code here
trials = []

for i in range(100): # outer loop
    n = 0
    while True: # inner loop
        n += 1
        die_1 = random.randint(1, 6)
        die_2 = random.randint(1, 6)
        if die_1 + die_2 in [7, 11]:
            break
    trials.append(n)
    
print('The average number of rolls to get a 7 or 11 is {}'.format(sum(trials) / len(trials)))

The average number of rolls to get a 7 or 11 is 4.58


### Problem 14: Simulating Craps - Doubling up or going broke?
<span style="color:green">Its possible to simulate someone playing craps until a certain criteria is met. Continually play the game of craps until the balance has doubled or until the player is out of money. Use the code from above but modify it in such a way that the print statements are removed and that the balance is updated each game. Keep track of the number of games played and print this number out with the ending balance.</span>

In [115]:
balance = 500
wager = 10
games = 0
# your code here

while 0 < balance < 1000:
    games += 1
    # Add in second stage of game
    die_1 = random.randint(1, 6)
    die_2 = random.randint(1, 6)
    total = die_1 + die_2

    if total in [2, 3, 12]:
        balance -= wager
    elif total in [7, 11]:
        balance += wager
    else:
        while True:
            die_1 = random.randint(1, 6)
            die_2 = random.randint(1, 6)
            new_total = die_1 + die_2
            if new_total == total:
                balance += wager
                break
            if new_total == 7:
                balance -= wager
                break
            
print('After {} games the balance is {}'.format(games, balance))

After 766 games the balance is 1000


### Problem 15:
<span style="color:green">Simulate the scenario in problem 14, 1000 times and determine what percentage of the time you would end up doubling your money. Use a print statement to write out results. Is craps a good investment? (this takes a while to simulate)</span>

In [119]:
n = 1000
doubling = 0
for i in range(1000):
    balance = 500
    wager = 10
    games = 0
    # your code here

    while 0 < balance < 1000:
        games += 1
        # Add in second stage of game
        die_1 = random.randint(1, 6)
        die_2 = random.randint(1, 6)
        total = die_1 + die_2

        if total in [2, 3, 12]:
            balance -= wager
        elif total in [7, 11]:
            balance += wager
        else:
            while True:
                die_1 = random.randint(1, 6)
                die_2 = random.randint(1, 6)
                new_total = die_1 + die_2
                if new_total == total:
                    balance += wager
                    break
                if new_total == 7:
                    balance -= wager
                    break
                    
    if balance == 1000:
        doubling += 1

pct_doubling = doubling / n * 100
        
print("After {} trials, money was doubled {} percent of the time".format(n, pct_doubling))

After 1000 trials, money was doubled 20.1 percent of the time


### Problem 16:
<span style="color:green">Use a list comprehension to create a list of the square root of the first 10 positive integers</span>

In [120]:
# your code here
[math.sqrt(x) for x in range(1, 11)]

[1.0,
 1.4142135623730951,
 1.7320508075688772,
 2.0,
 2.23606797749979,
 2.449489742783178,
 2.6457513110645907,
 2.8284271247461903,
 3.0,
 3.1622776601683795]

### Problem 17
<span style="color:green">Use a list comprehension to create a list of the area of a circle for circles with a radius equal to the first 10 postive integers.</span>

In [121]:
# your code here
[math.pi * r ** 2 for r in range(1, 11)]

[3.141592653589793,
 12.566370614359172,
 28.274333882308138,
 50.26548245743669,
 78.53981633974483,
 113.09733552923255,
 153.93804002589985,
 201.06192982974676,
 254.46900494077323,
 314.1592653589793]

### Problem 18
<span style="color:green">Create a list of strings. Then use a list comprehension to create a list that contains the length of each of these strings.</span>

In [122]:
# your code here
string_list = ['list', 'comprehensions', 'rock']
[len(word) for word in string_list]

[4, 14, 4]

### Problem 19
<span style="color:green">Create a list of strings. Then use a list comprehension to create a list that contains only the strings that have length greater than 5.</span>

In [123]:
# your code here
string_list = ['list', 'comprehensions', 'are', 'fantastic']
[word for word in string_list if len(word) > 5]

['comprehensions', 'fantastic']

### Problem 20
<span style="color:green">Create a long string. Use a list comprehension to create a list of only the vowels in that string.</span>

In [124]:
# your code here
long_string = 'apiunkwqenasvasdoiureh'
[letter for letter in long_string if letter in ['a', 'e', 'i', 'o', 'u']]

['a', 'i', 'u', 'e', 'a', 'a', 'o', 'i', 'u', 'e']

### Problem 21

<span style="color:green">Write a list comprehension that loops through all numbers between 1 and 100 and squares all those that end in 3 or 7 (hint: use modulus operator)</span>

In [125]:
# your code here
[i for i in range(100) if i % 10 in [3, 7]]

[3, 7, 13, 17, 23, 27, 33, 37, 43, 47, 53, 57, 63, 67, 73, 77, 83, 87, 93, 97]

### Problem 22

<span style="color:green">Rewrite the following code using the ternary syntax </span>

`>>> x = 100
    if x > 99:
         label = 'perfect'
    else:
         label = 'still need work to do'`

In [127]:
# your code here
x = 100
label = 'perfect' if x > 99 else 'still need work to do'

label

'perfect'

### Problem 23

<span style="color:green">Use the ternary syntax to square a number if its even and raise it to the third power if its odd </span>

In [128]:
# your code here
x = 19
result = x ** 2 if x % 2 == 0 else x ** 3

result

6859

### Problem 24

<span style="color:green">Write a list comprehension that uses nested for loops to get the total price of carpeting for all possible house size combinations where there is at least 800 square feet of house. Store values in total_price list</span>

In [10]:
height = range(10, 30, 5)
width = range(2,40,2)
price_per_square_foot = [1, 1.75, 2.5, 4.88]
total_price = []
# your code here
total_price = [h * w * p for h in height for w in width for p in price_per_square_foot if h * w > 800]

total_price

[850,
 1487.5,
 2125.0,
 4148.0,
 900,
 1575.0,
 2250.0,
 4392.0,
 950,
 1662.5,
 2375.0,
 4636.0]

### Problem 25

<span style="color:green">Write a for list that uses enumerate to iterate over the total_price list in problem 22 and prints out the message, "House number i costs x amount of dollars". Use the format method of str object to interpolate the variables in your message </span>

In [13]:
# your code here
for i, price in enumerate(total_price):
    print("House number {} costs {} amount of dollars".format(i, price))

House number 0 costs 850 amount of dollars
House number 1 costs 1487.5 amount of dollars
House number 2 costs 2125.0 amount of dollars
House number 3 costs 4148.0 amount of dollars
House number 4 costs 900 amount of dollars
House number 5 costs 1575.0 amount of dollars
House number 6 costs 2250.0 amount of dollars
House number 7 costs 4392.0 amount of dollars
House number 8 costs 950 amount of dollars
House number 9 costs 1662.5 amount of dollars
House number 10 costs 2375.0 amount of dollars
House number 11 costs 4636.0 amount of dollars


# 5. Functions

### Problem 1

<span style="color:green">Create three different variables with three different data types then use the `type` function to verify that the types are indeed different</span>

In [28]:
# your code here
a = True
b = 5
c = 7.4
type(a), type(b), type(c)

(bool, int, float)

In [24]:
pow(2,5)

32

In [26]:
code_string = '''
a = 5
b = 7
print(a + b)
'''

exec(code_string)

12


### Problem 2

<span style="color:green">The `eval` builtin function, takes a string as an argument and executes that string if it were Python code. Create a variable that stores a string that looks like Python code. Then use the `eval` function to execute this code string.</span>

In [34]:
# your code here
code_string = '5 + 5'

eval(code_string)

10

### Problem 3

<span style="color:green">Use the help or [read the official documentation](https://docs.python.org/3/library/functions.html) about how to use the `all` and `any` functions and use them correctly below.</span>

In [38]:
# your code here
bool_list = [True, False, 5 > 0]

print(all(bool_list), any(bool_list))

False True


### Problem 4

<span style="color:green">Use two consecutive print functions. Pass each print function three positional arguments and separate them with a period and end them with two new lines</span>

In [45]:
# your code here
print('Hello', 'World', '!', sep='.', end='\n\n')
print('Bye', 'World', '!', sep='.', end='\n\n')

Hello.World.!

Bye.World.!



### Problem 5

<span style="color:green">Create a function that finds the area of a circle and defaults the value of the radius to 0 if it's not given.</span>

In [52]:
# your code here
def area_circle(radius=0):
    return 3.14 * radius ** 2

area_circle()

0.0

### Problem 6

<span style="color:green">Create a function that takes an string argument and returns True or False whether the string is a palindrome (spelled the same forward and backwards)</span>

In [44]:
# your code here
def is_palindrome(word):
    return word == word[::-1]

is_palindrome('racecar')

True

### Problem 7

<span style="color:green">Create a function `concat_sort_list` that takes two lists and concatenates them together, sorts the list and then returns the sorted list.</span>

In [48]:
# your code here
def concat_sort_list(list_1, list_2):
    final_list = list_1 + list_2
    final_list.sort()
    return final_list

In [49]:
# test code with
concat_sort_list([4, 5, 2], [9, 0, -8]) == [-8, 0, 2, 4, 5, 9]

True

### Problem 8

<span style="color:green">Create an anonymous function that calculates the area of a circle. Store it to a variable and check that it works.</span>

In [53]:
# your code here
area_circle = lambda radius: 3.14 * radius ** 2

area_circle(5)

78.5

### Problem 9

<span style="color:green">Use the sorted function to sort a list of strings using the last letter. Use an anonymous function passed to the key argument.</span>

In [54]:
# your code here
string_list = ['banana', 'strawberry', 'orange', 'apple']
sorted(string_list, key = lambda word: word[-1])

['banana', 'orange', 'apple', 'strawberry']

### Problem 10

<span style="color:green">Create a list of integers and sort them by their last digit. Use an anonymous function passed to the key argument.</span>

In [55]:
# your code here
int_list = [109, 50, 434, 123131]
sorted(int_list, key = lambda x: x % 10)

[50, 123131, 434, 109]

### Problem 11: Advanced

<span style="color:green">Use the sorted function (with the `key` argument) to sort a list of strings by the third lowest alphabetical letter. Use a lambda expression to find this third lowest letter. Take the word 'python' for example. If you sort it by letter it becomes hnopty. The third lowest letter is therefore 'o'.</span>

In [2]:
# make a string list with each string having three or more characters
string_list = 'some string list with many different words'.split()
# your code here
sorted(string_list, key=lambda x: sorted(x)[2])

['different', 'string', 'many', 'some', 'words', 'list', 'with']

### Problem 12

<span style="color:green">Use the map function to square each number in a list. Do the same with a list comprehension</span>

In [58]:
# your code here
my_list = [9, 5, 1, 99]
list(map(lambda x: x ** 2, my_list))

[81, 25, 1, 9801]

In [59]:
[x ** 2 for x in my_list]

[81, 25, 1, 9801]

### Problem 13

<span style="color:green">Read the documentation on the **`filter`** builtin function. Use an anonymous function to filter out the even values from a range object of values 0 to 9</span>

In [60]:
# your code here
list(filter(lambda x: x % 2 == 1, range(10)))

[1, 3, 5, 7, 9]

### Problem 14

<span style="color:green">Read the documentation on the **`reduce`** function which is part of the functools standard library. Use an anonymous function to sum up all the values from 0 to 9</span>

In [61]:
from functools import reduce
# your code here
reduce(lambda x, y: x + y, range(10))

45

### Problem 15: Advanced

<span style="color:green">Use **`map, reduce and filter`** to take the numbers 0 to 9, square them, filter out the even numbers and multiply the remaining together. This can be done in one line</span>

In [145]:
# your code here
reduce(lambda x, y: x * y, filter(lambda x: x % 2 == 1, map(lambda x: x ** 2, range(10))))

893025

### Problem 16: Advanced

<span style="color:green">Use list comprehensions and the **`sum`** function to take the numbers 0 to 9, square the even numbers and sum them all up</span>

In [146]:
# your code here
sum([x ** 2 for x in range(10) if x % 2])

165

### Problem 17: Advanced
<span style="color:green">In this problem you will write a function to transpose a two dimensional matrix. Transposing a matrix means to make the rows become the columns and vice versa. Create your matrix by creating a list of lists. In this case just choose a simple matrix like a 2 by 3.</span>

In [63]:
# First lets pretty print our 3x2 matrix
matrix = [[1,2,3], [4,5,6]]
for row in matrix:
    print(row)

[1, 2, 3]
[4, 5, 6]


In [64]:
# need to convert this matrix to a 2 x 3 with rows and columns swapped
# assume each inner list is a row, the number of columns is the length of one inner list
def transpose(matrix):
    new_num_rows = len(matrix[0])
    new_num_cols = len(matrix)
    new_matrix = []
    for i in range(new_num_rows):
        new_matrix.append([])
        for j in range(new_num_cols):
            new_matrix[-1].append(matrix[j][i])
    return new_matrix  

In [65]:
for row in transpose(matrix):
    print(row)

[1, 4]
[2, 5]
[3, 6]


### Problem 18: Advanced
Create a tic-tac-toe function that takes four parameters  

1. **board**: a list of lists representing the game board  
1. **mark**: 'x' or 'o'  
1. **row**: Row that the mark will be played
1. **col**: Column that the mark will be played

Returns the updated board or the winner.  
**`tic_tac_toe(board, 'x', 2, 0)`** should place an 'x' on the last row in the first position

In [62]:
def tic_tac_toe(board, mark, row, col):
    '''
    This function accepts a 3x3 tic tac toe board and places a mark
    at its row and col
    It does some value checking of the parameters and makes sure no spot is already taken
    It returns the board plus a message if there is a winner or a tie (cats game)
    
    Parameters
    -----
    board: 3x3 list of lists
    mark: 'x' or 'o'
    row, col: integer between 0 and 2 
    '''
    if mark not in ['x', 'o']:
        raise ValueError('Mark must be "x" or "y"')
    if row not in range(3) or col not in range(3):
        raise ValueError('row and col must be between 0 and 2')
    if board[row][col] != None:
        raise ValueError('That position is already taken')
        
    board[row][col] = mark
    winner = False
    
    for row in board:
        print(row)
    
    
    diag1 = board[0][0] == board[1][1] == board[2][2] and board[0][0] is not None
    diag2 = board[0][2] == board[1][1] == board[2][0] and board[1][1] is not None
    
    for i in range(3):
        row_check = board[i][0] == board[i][1] == board[i][2] and board[i][0] is not None
        col_check = board[0][i] == board[1][i] == board[2][i] and board[0][i] is not None
        if row_check or col_check or diag1 or diag2:
            winner = True
            
    if winner:        
        print("Winner is " + mark)
        
    if len([1 for r in board for c in r if c is not None]) == 9:
        print("Cats Game")
    
    return board

In [66]:
board = [[None, None, None], [None, None, None], [None, None, None]]

In [67]:
# start playing!
board = tic_tac_toe(board, 'x', 0, 2)

[None, None, 'x']
[None, None, None]
[None, None, None]


In [68]:
# can't play on someone else's spot
board = tic_tac_toe(board, 'o', 0, 2)

ValueError: That position is already taken

In [69]:
board = tic_tac_toe(board, 'o', 1, 2)

[None, None, 'x']
[None, None, 'o']
[None, None, None]


In [70]:
board = tic_tac_toe(board, 'x', 1, 1)

[None, None, 'x']
[None, 'x', 'o']
[None, None, None]


In [71]:
board = tic_tac_toe(board, 'o', 2, 1)

[None, None, 'x']
[None, 'x', 'o']
[None, 'o', None]


In [72]:
board = tic_tac_toe(board, 'x', 2, 0)

[None, None, 'x']
[None, 'x', 'o']
['x', 'o', None]
Winner is x


### Problem 19
<span style="color:green">Write a function that returns the roll of two dice. Use the random.randint function</span>

In [75]:
import random
# your code here
def dice_roll():
    return random.randint(1, 6), random.randint(1, 6)

dice_roll()

(5, 6)

### Problem 20
<span style="color:green">You will write a function that plays a simple dice game. The game works as follows: You roll two dice and record the sum. You keep rolling the two dice until you re-roll your original total. Return the number of rolls it took to re-roll the original total. Do not simulate the original roll. Instead use an argument to pass the function a number between 2 and 12. Use the function you created in problem 19 to roll the dice. </span>

In [81]:
def simple_dice_game(original_roll):
    # your code here
    n = 0
    while True:
        n += 1
        if sum(dice_roll()) == original_roll:
            break
    return n

simple_dice_game(5)

11

### Problem 21: Advanced
<span style="color:green">This is a classic interview problem. Write a function that finds the degree difference between the minute and second on a clock. The function will accept two arguments, hour (between 1 and 12) and minute (between 0 and 59). Note: this should the absolute difference, so 180 degrees would be the absolute max possible value. Test your function with some times and make sure it makes sense. </span>

In [82]:
def degree_diff(hour, minute):
    hour_degree = ((hour % 12) + minute / 60) * 30 
    minute_degree = minute * 6
    return min(abs(hour_degree - minute_degree), 360 - abs(hour_degree - minute_degree))

In [122]:
degree_diff(1, 15)

52.5

# 6. Tuples, Sets and Dictionaries

### Problem 1

<span style="color:green">Find all the public tuple methods. There was a similar example done in the lists notebook if you forgot.</span>

In [123]:
# your code here
[method for method in dir(tuple()) if method[0] != '_']

['count', 'index']

### Problem 2

<span style="color:green">Define a set and think of a function that will output the number of elements in that set.</span>

In [124]:
# your code here
a_set = {4, 5, 10}
len(a_set)

3

### Problem 3

<span style="color:green">Create a set of all the even numbers from 0 to 100. Then create another set of every number divisible by 7 from 0 to 100. Find the intersection of those sets.</span>

In [128]:
# your code here
a = set(range(0, 101, 2))
b = set(range(0, 101, 7))

a & b

{0, 14, 28, 42, 56, 70, 84, 98}

### Problem 4

<span style="color:green">Find all the public tuple methods that are not in common with set methods.</span>

In [131]:
# your code here
tuple_methods = set(dir(tuple))
set_methods = set(dir(set))
tuple_unique = tuple_methods - set_methods
[method for method in tuple_unique if method[0] != '_']

['index', 'count']

### Problem 5: Advanced
<span style="color:green">Create a function to make a list of all the Fibonacci numbers less than 1000 and another function to make a list of all the prime numbers less than a 1000. Use a set to find the numbers that are in both groups</span>

In [134]:
# your code here
fib_list = [1, 1]
while  fib_list[-1] + fib_list[-2] < 1000:
    fib_list.append(fib_list[-2] + fib_list[-1])
    
prime_list = [1, 2]
from math import sqrt
for i in range(3, 1000):
    is_prime = True
    for j in range(2, int(sqrt(i) + 1)):
        if i % j == 0:
            is_prime = False
            break
    if is_prime:
        prime_list.append(i)

In [150]:
set(fib_list) & set(prime_list)

{1, 2, 3, 5, 13, 89, 233}

### Problem 6

<span style="color:green">Are dictionaries mutable or immutable objects?</span>

Mutable. We can add new key:value pairs, remove key:value pairs or change the value of one of the keys

### Problem 7

<span style="color:green">Write a function that emulates the `get` dictionary method. The function will consist of three arguments: the dictionary, the key to lookup and the default value to return if the key is not in the dictionary. Test your function with the provided code</span>

In [136]:
def dict_get(dictionary, key, default):
    if key in dictionary:
        return dictionary[key]
    return default

In [137]:
# test code 
test_dict = {'Sally': [87, 76, 65], 'Jane' : [45, 98, 77], 'Adeline' : [65, 22, 10]}
key = 'jane'
default = [0, 0, 0]

print(dict_get(test_dict, key, default))

key = 'Jane'
print(dict_get(test_dict, key, default))

[0, 0, 0]
[45, 98, 77]


### Problem 8

<span style="color:green">Create a function that accepts two parameters, a dictionary, and a string. If that string is in the dictionary, increment the value by 1. If not, create a new record in the dictionary with the string as a key and 0 as its value. Test your function with provided code.</span>

In [138]:
def add_to_dict(dictionary, key):
    if key in dictionary:
        dictionary[key] += 1
    else:
        dictionary[key] = 0

In [142]:
# test code
test_dict = {'Houston': 1, 'New Orleans':2}

add_to_dict(test_dict, 'New York')

add_to_dict(test_dict, 'New Orleans')

test_dict

{'Houston': 1, 'New Orleans': 3, 'New York': 0}

### Problem 9

<span style="color:green">Create a new dictionary through a dictionary comp that gets the average score for each student</span>

In [143]:
students = {'Sally': [87, 76, 65], 'Jane' : [45, 98, 77], 'Adeline' : [65, 22, 10]}
# your code here
avg = {student : sum(scores) / len(scores) for student, scores in students.items()}
avg

{'Adeline': 32.333333333333336, 'Jane': 73.33333333333333, 'Sally': 76.0}

### Problem 10

<span style="color:green">Create a dictionary through a dictionary comp that has the student name as the key and the minimum grade as value but for only students that have an 'e' in their name.</span>

In [144]:
# your code here
low_e = {student : min(scores) for student, scores in students.items() if 'e' in student}
low_e

{'Adeline': 10, 'Jane': 45}

### Problem 11

<span style="color:green">Use a dictionary comp that loops through the numbers 0 to 10 and uses the integer as the key and a list with its squared and cubed value as its value.</span>

In [145]:
# your code here
{n:[n ** 2, n ** 3] for n in range(11)}

{0: [0, 0],
 1: [1, 1],
 2: [4, 8],
 3: [9, 27],
 4: [16, 64],
 5: [25, 125],
 6: [36, 216],
 7: [49, 343],
 8: [64, 512],
 9: [81, 729],
 10: [100, 1000]}

### Problem 12: Advanced

<span style="color:green">Iterate through each student and drop their lowest test grade. Replace it with 100.</span>

In [160]:
# your code here
for student, scores in students.items():
    scores[scores.index(min(scores))] = 100

In [161]:
students

{'Adeline': [65, 22, 100], 'Jane': [100, 98, 77], 'Sally': [87, 76, 100]}